### PR TITLE
Fixes #23013: Avoid denormalization refreshes for location assignment when not needed

### LINKED ISSUE #23013: Saving a Location propagates its site assignment to all descendant objects even when the site has not changed
### NetBox Edition

NetBox Community

### NetBox Version

v4.6.8

### Python Version

3.14

### Steps to Reproduce

1. Create a site "Site 1" (slug `site-1`).
2. Create a location "Location 1" (slug `location-1`) assigned to Site 1.
3. Create a device type, role, and manufacturer as needed, then create a device "Device 1" assigned to Site 1 and Location 1.
4. Add several interfaces to Device 1 (any number; the effect scales with the count).
5. Enable query logging on the PostgreSQL server, e.g. by setting `log_statement = 'all'` in `postgresql.conf` and reloading, so that the statements issued by the following request can be observed.
6. Update the location's description via the REST API, leaving its site assignment untouched:

   ```
   PATCH /api/dcim/locations/1/ HTTP/1.1
   Host: netbox.example.com
   Authorization: Token $TOKEN
   Content-Type: application/json
   
   {"description": "Updated description"}
   ```
7. Inspect the PostgreSQL query log for the statements issued while handling the request.

The same behavior occurs for any write to a Location which leaves `site` unchanged: editing it in the UI, assigning a tag, or setting a custom field value via `PATCH {"custom_fields": {...}}`.

### Expected Behavior

The `site` assignment of the location was not modified, so no descendant object needs its denormalized site reference rewritten. NetBox should issue only the statements needed to update the Location row itself (plus its changelog and search cache entries).

### Observed Behavior

NetBox issues a mass `UPDATE` against fourteen additional tables, rewriting every descendant location, rack, device, power panel, cable termination, and device component in the location's subtree — with the value they already hold:

```sql
UPDATE "dcim_location"       SET "site_id" = 1 WHERE <descendants of location 1>;
UPDATE "dcim_rack"           SET "site_id" = 1 WHERE "location_id" IN (1);
UPDATE "dcim_device"         SET "site_id" = 1 WHERE "location_id" IN (1);
UPDATE "dcim_powerpanel"     SET "site_id" = 1 WHERE "location_id" IN (1);
UPDATE "dcim_cabletermination" SET "_site_id" = 1 WHERE "_location_id" IN (1);
UPDATE "dcim_consoleport"    SET "_site_id" = 1 WHERE "device_id" IN (SELECT id FROM dcim_device WHERE location_id IN (1));
-- ...and likewise for consoleserverport, devicebay, frontport, interface,
--    inventoryitem, modulebay, poweroutlet, powerport, rearport
```

Because PostgreSQL writes a new tuple version for every row an `UPDATE` matches — whether or not the value actually changes — each of these statements incurs the full cost of a rewrite: new row versions, index maintenance, WAL, and subsequent autovacuum work. Every affected row is also row-locked for the remainder of the request transaction (REST API writes are wrapped in `transaction.atomic()`), blocking concurrent writers to those tables.

On a large installation this makes a trivial Location edit extremely expensive. The cost scales with the number of *device components* beneath the location, not the number of locations. A user with only 222 locations, but a large device fleet beneath the location being edited, reported that a `PATCH` setting a single custom field value on one location ran for over ten minutes before their proxy timed out; `pg_stat_activity` showed the request parked on the `dcim_interface` update. The same edit performed through the UI is affected identically.

### Suspected Cause

`handle_location_site_change()` in `netbox/dcim/signals.py` (lines 93-119) runs its propagation block unconditionally on every `post_save` for an existing Location:

```python
@receiver(post_save, sender=Location)
def handle_location_site_change(instance, created, using=None, **kwargs):
    if created:
        return
    with transaction.atomic(using=using, savepoint=False):
        instance.get_descendants().using(using).update(site_id=instance.site_id)
        locations = list(instance.get_descendants(include_self=True).using(using).values_list('pk', flat=True))
        Rack.objects.using(using).filter(location__in=locations).update(site_id=instance.site_id)
        Device.objects.using(using).filter(location__in=locations).update(site_id=instance.site_id)
        PowerPanel.objects.using(using).filter(location__in=locations).update(site_id=instance.site_id)
        CableTermination.objects.using(using).filter(_location__in=locations).update(_site_id=instance.site_id)
        for model in COMPONENT_MODELS:
            model.objects.using(using).filter(device__location__in=locations).update(_site_id=instance.site_id)
        ...
```

`COMPONENT_MODELS` (`netbox/dcim/signals.py:43`) comprises ten models: ConsolePort, ConsoleServerPort, DeviceBay, FrontPort, Interface, InventoryItem, ModuleBay, PowerOutlet, PowerPort, and RearPort.

Every value written by this block is derived solely from `instance.site_id`, so when that field is unchanged the entire block is a no-op in terms of resulting data — but not in terms of cost.

The machinery needed to detect this case already exists. netbox-community/netbox#22810 added `cache_presave_scope_fields()` (`netbox/dcim/signals.py:61-89`), which stashes the pre-save `site_id` on the instance under `_presave_scope_fields`, and applied the resulting guard in two places:

* `sync_cached_scope_fields()` (`netbox/dcim/signals.py:393-403`), which returns early when the scope fields are unchanged, and
* the *trailing* block of `handle_location_site_change()` (`netbox/dcim/signals.py:127`), which skips the Prefix/Cluster/WirelessLAN/CircuitTermination rebuild on the same condition.

The leading block quoted above was not covered, so on v4.6.7 and later the cheap portion of the work is skipped while the expensive portion still runs on every save.

### Proposed Fix

Apply the existing guard to the whole of `handle_location_site_change()` rather than only to its trailing block: after the `created` check, read `_presave_scope_fields` and return early when the stash is present and `prev['site_id'] == instance.site_id`. The trailing block's own check then becomes redundant and can be removed.

The stash is deliberately absent for saves made outside a transaction (see the docstring of `cache_presave_scope_fields()`), in which case the handler must continue to propagate unconditionally — matching the behavior already established in `sync_cached_scope_fields()`.

This preserves correctness for every case in which the propagation is meaningful, since all of the values written are computed from `instance.site_id`. It does forgo the incidental repair of rows holding stale cached values on an unrelated save, which is the same tradeoff netbox-community/netbox#22810 already accepted for `sync_cached_scope_fields()`.

Suggested regression test: save an existing Location without modifying its site, and assert with `assertNumQueries()` (or by capturing `connection.queries`) that no `UPDATE` is issued against `dcim_interface` or the other component tables.

### PR DESCRIPTION
**Pull request by jeremystretch**
## Summary

`handle_location_site_change()`, `handle_rack_site_change()`, and `handle_device_site_change()` each rewrote every descendant row beneath the saved object on **every** save, whether or not the assignments they propagate had actually changed. A `PATCH` setting a single custom field on a Location — or a UI edit changing only a description on a Location, Rack, or Device — issued mass `UPDATE`s against as many as fourteen tables.

This is not merely wasted work. PostgreSQL writes a new tuple version for every row an `UPDATE` matches, whether or not the value differs, so each statement incurs the full cost of a rewrite: new row versions, index maintenance, WAL, and subsequent autovacuum. Every affected row is also row-locked for the remainder of the request transaction (REST API writes are wrapped in `transaction.atomic()`), blocking concurrent writers to those tables.

The cost scales with the number of *device components* beneath the saved object, not the number of locations. The reporter in #23013 has only 222 locations but a large device fleet, and a `PATCH` setting one custom field ran for over ten minutes before their proxy timed out, parked on the `dcim_interface` update.

## History

This is not a recent regression. The unconditional rewrite entered in **v2.11.0** and became expensive in **v4.4.9**.

| Release | Date | Change |
|---|---|---|
| **v2.11.0** | 2021-04-16 | `5406e8e2e` "Optimize background site/location updates" — the origin |
| v3.5.4 | 2023-06-20 | `0b2162569` adds `CableTermination` to the Location sweep |
| v4.3.6 | 2025-08-12 | #19977 adds the ten-model `COMPONENT_MODELS` sweep, to the Device handler only |
| **v4.4.9** | 2025-12-23 | #20875 extends that sweep to the Location and Rack handlers — the amplification |
| v4.6.7 | 2026-07-30 | #22810 adds the skip guard, but only to `sync_cached_scope_fields()` and the trailing scoped-objects block |

Before v2.11.0 these handlers looped per object and filtered with `.exclude(site=instance.site)`, so a save which moved nothing touched nothing. `5406e8e2e` replaced those loops with blanket queryset `update()` calls:

```diff
-        for rack in Rack.objects.filter(location=instance).exclude(site=instance.site):
-            rack.site = instance.site
-            rack.save()
+        Rack.objects.filter(location__in=locations).update(site=instance.site)
```

That was a real optimization for an actual move — one statement instead of N saves — but it dropped the row-level predicate which made no-op saves free, in `handle_rack_site_change()` as well as `handle_location_site_change()`. **This PR restores the pre-v2.11 semantics of touching only what needs touching, while keeping the set-based queries `5406e8e2e` introduced.** It moves the predicate from per-row to per-save, which is where it belongs: the fields are homogeneous across the whole subtree, so one comparison on the saved object decides it for every descendant.

What made the latent inefficiency a production problem was #20875 in v4.4.9. Before it, a Location save rewrote rows proportional to the locations, racks, devices, and cable terminations in the subtree. After it, the sweep is keyed on `device__location__in=locations` across ten component tables, so the row count becomes every component of every device beneath the Location — the difference between a slow request and the ten-minute timeout in #23013.

The heavy sweep is byte-identical between v4.6.6 and v4.6.8, so v4.6.7's #22810 neither introduced nor worsened this; it added the very mechanism this PR now applies to the rest of the work.

## Changes

`STASHED_SCOPE_FIELDS` declares, per model, the scope-relevant fields read from the database immediately before each save. Each entry covers the trigger fields of every handler that reads it — the propagation handlers below and `sync_cached_scope_fields()` — so `Site` is listed for the latter alone, nothing propagating a Site's own scope downwards:

| Model | Stashed fields |
|---|---|
| `Site` | `region_id`, `group_id` |
| `Location` | `site_id` |
| `Rack` | `site_id`, `location_id` |
| `Device` | `site_id`, `location_id`, `rack_id` |

`cache_presave_scope_fields()` — added in #22810 for Site and Location — now fires for Rack and Device as well, taking its field list from that map. `_scope_fields_unchanged()` is the single definition of "this save propagates nothing", and all four handlers return early on it.

`sync_cached_scope_fields()` previously carried its own hand-written copy of that comparison; it now uses the shared helper, so the guard cannot drift between call sites as models are added. `handle_rack_site_change()` and `handle_device_site_change()` were converted to early returns to match `handle_location_site_change()`.

Everything each handler writes is derived solely from the fields listed above, so skipping is safe when none of them changed. For Location specifically, `Location.clean()` requires a parent location to belong to the same site, so a subtree's site is homogeneous and `site_id` is the only field that can make propagation necessary.

All four handlers also return early on a **raw save**. `raw=True` is set only by Django's `loaddata` pathway, whose fixture already carries the denormalized values for every object it loads, so the propagation would rewrite each matched row with what it already holds. `netbox.denormalized.update_denormalized_fields()` has always returned early on `raw` for this reason; the two mechanisms now agree about what a raw save means. Loading a `dumpdata` fixture previously issued roughly fifteen queryset `UPDATE`s per Location, eleven per Rack, and ten per Device, every one of them a provable no-op.

`handle_rack_site_change()` and `handle_device_site_change()` now wrap their propagation in `transaction.atomic(using=using, savepoint=False)`, matching `handle_location_site_change()` and `sync_cached_scope_fields()`. Inside a request transaction this costs nothing — no savepoint is taken — while in autocommit (a background job, a management command, `nbshell`) it makes the fan-out all-or-nothing. Previously a Rack move there issued its `dcim_device` update plus ten component-table updates as eleven separate transactions, so a mid-sequence failure could leave the Rack at its new site and location with some of its devices' components still caching the old ones. That gap predates this PR, but the skip guard is what makes the resulting drift permanent rather than self-healing on the next save, so it is closed here.

### Note for reviewers: `.order_by()` on the pre-save read

The pre-save read now clears the model's default ordering. `Rack.Meta.ordering` is `('site', 'location', 'name', 'pk')`; ordering by those foreign keys resolves through the related models' own `Meta.ordering` and adds a `LEFT OUTER JOIN` on the nullable `location`, which PostgreSQL refuses to lock:

```
NotSupportedError: FOR NO KEY UPDATE cannot be applied to the nullable side of an outer join
```

Site and Location never hit this because their ordering is local columns only. The query selects a single row by primary key, so ordering was contributing nothing but joins.

### Note for reviewers: cost of the Device guard

Every in-transaction Device save now issues one extra `SELECT ... FOR NO KEY UPDATE` against its own row. That row is locked by the save's own `UPDATE` moments later, so this introduces no new lock ordering and no additional deadlock surface — but device saves are a hot write path, so it is a real if small per-save cost, bought against skipping ten component-table sweeps.

## Behavioral tradeoff

A save which changes none of the propagated fields no longer incidentally repairs descendant rows holding stale cached values. This is the same tradeoff #22810 already accepted for `sync_cached_scope_fields()`.

Skipping raw saves means `loaddata` no longer recomputes cached scope fields as a side effect. That is safe for the `dumpdata`/`loaddata` round-trip which sets `raw=True`, since the fixture carries those columns; a hand-written fixture omitting them would load with them unpopulated — as it already would for every other denormalized field, `netbox.denormalized` having always skipped raw.

The stale-value guard applies only to saves made inside a transaction. `cache_presave_scope_fields()` deliberately takes no stash in autocommit, where the pre-save read and the save's `UPDATE` would run in separate transactions and the comparison could race a concurrent save; a missing stash means "the values may have changed" and propagation runs unconditionally.

## Tests

`ScopePropagationCaptureMixin` captures the `UPDATE` statements a save issues against the tables its handler propagates to. `dcim_cabletermination` is excluded throughout — the denormalized-field registry (`netbox.denormalized`) rewrites it on Location, Rack, and Device saves alike, so it cannot distinguish a propagation from a plain save — as is the saved object's own table, which carries the save's own `UPDATE`.

Each handler gets a skip test and a positive control, so the skip tests cannot pass vacuously if the `UPDATE`s stop being issued for an unrelated reason or a table is renamed:

- **Location** — `test_unchanged_site_skips_propagation`, `test_changed_site_propagates`
- **Rack** — `test_unchanged_scope_skips_propagation`, `test_changed_site_propagates`, `test_changed_location_propagates` (a move within one site, which only the second guarded field catches)
- **Device** — `test_unchanged_scope_skips_propagation`, `test_changed_site_propagates`, `test_changed_rack_propagates` (the rack is deliberately left without a location so `Device.save()` inherits none, making `rack_id` the only field that moves)
- **Raw saves** — `test_raw_save_skips_propagation` on each of the three propagation handlers and `test_raw_save_skips_resync` on `sync_cached_scope_fields()`. Each makes a *real* scope change before saving raw, so the assertion cannot hold merely because there was nothing to propagate. `test_raw_save_does_not_reuse_a_previous_saves_stash` covers the stash being cleared on a raw save, keeping the invariant that a stash never outlives the save it was taken for.
- **Autocommit** — `LocationSiteChangeAutocommitTestCase` covers the no-stash path, where propagation must remain unconditional
- **Registration** — `StashedScopeFieldsRegistrationTestCase` asserts the map's keys match the models whose receiver is connected and that every entry's fields resolve on a real save, so an entry and its receiver cannot drift apart

Verified against the unfixed handlers: a description-only Location save rewrote `dcim_rack`, `dcim_device`, `dcim_powerpanel` and all ten `COMPONENT_MODELS` tables, and both Rack and Device skip tests fail without their guards. All four raw-save tests likewise fail with the `raw` guard neutered. All positive controls pass in both states.

`python manage.py test dcim ipam circuits virtualization wireless` — 6684 tests, `OK (skipped=5)`.

### CODE REVIEW

File: netbox/dcim/signals.py [hunk 1]
```diff
@@ -90,8 +114,32 @@ def cache_presave_scope_fields(instance, raw=False, using=None, **kwargs):
...
+    propagation or rebuild can be skipped in its entirety.
+
+    A missing stash means the values may have changed — cache_presave_scope_fields()
+    deliberately takes none for a raw save, for a new instance, or outside a transaction —
+    so the caller must do its work unconditionally.
+    """
+    prev = getattr(instance, '_presave_scope_fields', None)
+    if prev is None:
+        return False
+    return all(value == getattr(instance, field) for field, value in prev.items())
```
Comment (pheus) [inline netbox/dcim/signals.py]: Could we account for `update_fields` here? With a stale instance, `save(update_fields={'description'})` leaves the database scope unchanged, but this comparison can still see older in-memory scope values and cause the handlers to propagate them to descendants or components.

Could omitted scope fields be taken from the stashed values, with a regression test covering a stale partial save?

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks for the thorough work on this. The broader propagation guard and the additional test coverage look good overall.

I found one correctness edge case around partial saves: `update_fields` can leave stale scope values on the instance, which may then be propagated even though those fields were not written. I've left an inline comment with the details. I'm requesting changes for this case only. Once addressed, the overall approach looks good to me.

File: netbox/dcim/signals.py (hunk 1 above)
Comment (jeremystretch) [reply netbox/dcim/signals.py]: Good catch, thanks.

**Review by pheus — APPROVED** [review]
Thanks for addressing the `update_fields` edge case. The latest revision looks solid to me!
